File: 프로그래머스/3/43238. 입국심사/입국심사.py
def solution(n, times):
    answer = 0
    
    # 비어 있는 심사대에서 심사 받되, 더 빨리 끝나는 심사대가 있다면 기다림
    # 모든 사람이 심사를 받는데 걸리는 총 시간 최소화
    # 파라메트릭 서치 문제 -> 시간을 기준으로
    
    left = 1
    right = max(times) * n
    
    while (left <= right):
        mid = (left + right) // 2
        
        count = 0
        for time in times:
            # mid 시간 안에 몇 명을 심사할 수 있는지 계산
            count += mid // time
        
        if count >= n:
            # 모두 심사할 수 있음
            answer = mid
            right = mid - 1
        
        elif count < n:
            # 부족함
            left = mid + 1

        
        
    
    
            
    
    
    return answer

File: 프로그래머스/3/43238. 입국심사/입국심사.js
function solution(n, times) {
    var answer = 0;
    
    // 시간을 기준으로 탐색하는 이분탐색
    // 최대 시간은 n * 1 000 000 000 (심사대가 한 개인 경우)
    let left = 1
    let right = n * 1000000000
    
    while (left <= right) {
        const mid = Math.floor((left + right) / 2)
        
        let total = 0
        for (const time of times) {
            // 각 심사대마다 mid 시간 동안 심사할 수 있는 사람 수
            total += Math.floor(mid / time)
        }
        
        if (total >= n) {
            answer = mid
            right = mid - 1
        }
        else {
            left = mid + 1
        }
    }
    
    
    
    return answer;
}